### ISSUE TITLE
Interpolation of symmetric tensors

### ISSUE THREAD

**Issue by gslndlb** [gridap/Gridap.jl#908]
I would like to use symmetric tensors in Gridap.

    using Gridap, Gridap.TensorValues

    𝒯 = CartesianDiscreteModel((0,2,0,1),(40,20))
    Ω = Interior(𝒯)
    dΩ = Measure(Ω,2)

    Γ = Boundary(𝒯)
    dΓ = Measure(Γ, 2)
    
    refₙ = ReferenceFE(lagrangian, SymTensorValue{2, Float64, 3}, 1)
    testₙ = TestFESpace(𝒯, refₙ)
    N = TransientTrialFESpace(testₙ)

    xx = 0:0.05:1.95
    yy = 0:0.05:.95

    qxx = SymTensorValue(1/2., 0., -1/2.)
    qi = [qxx for i = 1:length(xx), j = 1:length(yy)]
    q₀ = interpolate_everywhere(qi, N(0.0))

It fails at the definition of the FESpace, so I followed an advice found on the gitter and pasted all functions from [here](https://github.com/gridap/GridapHybrid.jl/blob/main/src/tmp_fixes_symmetric_valued_lagrangian_reffes.jl) at the beginning of my code:

Now, the definition of the FESpace works but it fails at interpolation with
`ERROR: dof ids either positive or negative, not zero`

The Julia file to reproduce the bug is quite long so I attach it here (as .txt because Github doesn't accept .jl)
[minimal_SymTensorValue.jl.txt](https://github.com/gridap/Gridap.jl/files/11726751/minimal_SymTensorValue.jl.txt)

**Comment by JordiManyer** [comment]
@gslndlb I have not had a change to look at your fixes yet, but the error is quite indicative. I believe you are setting some of your DoFs with id 0, which creates an error. In Gridap, free dofs have positive (1,2,3...) ids, while constrained dofs (by dirichlet bcs, for instance) have negative ids (-1,-2,...).

**Comment by gslndlb** [comment]
@JordiManyer thanks for your reply.
The reason I want to do this is because I'm working on equations from nematic liquid crystals, in which the unknown is the nematic tensor. It is a field of a second-order tensor, symmetric and traceless. So I could use 3-components vectors and overwrite standard tensor functions (trace, contraction and divergence) but I thought symmetric tensors were implemented so it would have saved me some time. A workaround is to use non-symmetric tensors, at the cost of performance (not such a problem in 2D because it only adds one component, but in 3D it adds 3 components...).

So if I understood well, the problem is that I need comp_to_dof[3] to be t21 in order other standard tensor-related functions to work (Trace, contractions...), but when building the Finite element space, it is expected that every component corresponds to an actual dof, which t21 is not in the case of symmetric tensors.

I am discovering Gridap, Julia and Finite elements at the same time, so I'm not sure I have the ability to do it. How complicated do you think it is? In particular I'm more used to object-oriented languages (Python and C++) so I'm a bit lost in the structure of the code, to understand which function applies to what.

**Comment by amartinhuertas** [comment]
@gslndlb ... I have succesfully used symmetric tensor fields in Gridap in the past. 

However, I faced a set of issues while working with the current implementation of symmetric tensors in Gridap's main branch; not sure if the issues I found are equivalent as the ones you found. In any case, I gathered all fixes in the following file which is yet to be merged with Gridap:

https://github.com/gridap/GridapHybrid.jl/blob/main/src/tmp_fixes_symmetric_valued_lagrangian_reffes.jl

just in case you might want to give these fixes a try  ...

**Comment by gslndlb** [comment]
@amartinhuertas Yes I included your functions in my Julia code (attached .jl file). I didn't include them in my post because it creates a very long code.

Without your functions I cannot build the FESpace. When I add your functions I can build the space but not interpolate.

**Comment by amartinhuertas** [comment]
> @amartinhuertas Yes I included your functions in my Julia code (attached .jl file). I didn't include them in my post because it creates a very long code.

Ok, I overlooked your first post, sorry ...

**Comment by amartinhuertas** [comment]
just in case it helps, the driver program that leverages symmetric tensors i available here:

https://github.com/gridap/GridapHybrid.jl/blob/main/test/StressAssistedHenckyHDGTests.jl

**Comment by JordiManyer** [comment]
> So if I understood well, the problem is that I need comp_to_dof[3] to be t21 in order other standard tensor-related functions to work (Trace, contractions...), but when building the Finite element space, it is expected that every component corresponds to an actual dof, which t21 is not in the case of symmetric tensors.

Yes, I believe that is exactly the problem. And conceptually it makes sense that every component corresponds to a single DoF. That is why in most of the fixes we are simply replacing `comp_to_dofs[comp]` by `comp_to_dofs.data[comp]`, which accesses the three components inside the tensor. 

I think that for this to work perfectly and without quirks, we should somehow add an additional interface that is capable of distinguish between the TensorValue (4 components) and the real info stored inside (3 dofs). But that might require some time to think trough, make sure we are not missing anything. 

In the meantime (and going back to my initial question) I think you might find it easier to do something like this: 

```julia
function Π(f::CellField)
  return Π_op∘f
end

function Π_op(x::VectorValue{3})
  return SymTensorValue(x.data)
end

𝒯 = CartesianDiscreteModel((0,2,0,1),(40,20))
Ω = Interior(𝒯)
dΩ = Measure(Ω,2)

Γ = Boundary(𝒯)
dΓ = Measure(Γ, 2)

refₙ = ReferenceFE(lagrangian, VectorValue{3,Float64}, 1)
testₙ = TestFESpace(𝒯, refₙ)
N = TransientTrialFESpace(testₙ)

xx = 0:0.05:1.95
yy = 0:0.05:.95
qxx = VectorValue(1/2., 0., -1/2.)
qi = [qxx for i = 1:length(xx), j = 1:length(yy)]
q₀ = interpolate_everywhere(qi, N(0.0))


a(u,v) = ∫(Π(u) ⊙ Π(v))*dΩ
A = assemble_matrix(a, N, N)
```

Conceptually, this is what I'm doing: You work with 3-component vector-valued spaces (which gives you the number of dofs that you want). This spaces are mature and will give you everything you want. On top of that, you add a "projection" function `Π` that takes your 3-component vectors and produces `SymTensorValues` out of them. This way, you can work with `Π(u)` and do everything you wanted to do. Gridap will handle the conversion behind the scenes. Some things might be less efficient (maybe you want to do your own spetialized gradient function, for instance) , but I believe everything can be done. I can give you support if you need anything specific (like how to setup your weak form, etc...). 

What do you think @amartinhuertas, am I missing anything important?

**Comment by JordiManyer** [comment]
Just for reference, the fix needed for `Gridap.ReferenceFEs._generate_face_own_dofs` was 

```julia
function Gridap.ReferenceFEs._generate_face_own_dofs(
  face_own_nodes, 
  node_and_comp_to_dof::AbstractVector{<:Gridap.TensorValues.SymTensorValue}
  )
  faces = 1:length(face_own_nodes)
  T = eltype(node_and_comp_to_dof)
  comps = 1:num_components(T)
  face_own_dofs = [Int[] for i in faces]
  for face in faces
    nodes = face_own_nodes[face]
    # Node major
    for comp in comps
      for node in nodes
        comp_to_dofs = node_and_comp_to_dof[node]
        dof = comp_to_dofs.data[comp]
        push!(face_own_dofs[face],dof)
      end
    end
  end

  face_own_dofs
end
```
It does fix some things, but there are still more places inside the code where you would have to do a similar thing.... For instance, `get_cell_dof_ids` still gives the wrong answer...

**Comment by gslndlb** [comment]
Thank you for your answers! I followed the idea of working with Vectors instead of tensors and to use a projection.
I used a slightly different approach which is to decompose on a basis of 2x2 tensor that respect the constraints of my unknown tensor (symmetric and traceless). My new unkowns are the coefficients of the decomposition on this basis. This is more flexible than using SymTensorValue, and removes 2 components (one for the tensor because of the traceless constraint and removes the Lagrange multiplier that used to impose the traceless constraint).

If the basis is orthonormal, most terms in the weak form can be expressed even without projection. Some specific terms need a projection.

I think it's generally a good idea to use a relevant basis rather than implementing all possible constraints (symmetric, traceless and others that one could think of). It will be tedious to express the basis in dimension d >= 4 but this seems a niche situation.
